File: Python/dynamic_programming/Climbing_Stairs.py
from typing import List

class Solution:
    def climbStairs(self, n: int) -> int:
        """
        Number of distinct ways to climb to the top using DP tabulation.
        Recurrence: dp[i] = dp[i-1] + dp[i-2]
        Base Cases:
            dp[0] = 1  # one way to stay at the ground (do nothing)
            dp[1] = 1  # one way to climb one step
        """
        if n <= 1:
            return 1
        
        dp = [0] * (n + 1)
        dp[0], dp[1] = 1, 1
        
        for i in range(2, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        
        return dp[n]


# -------------------------------
# Test Cases
# -------------------------------
if __name__ == "__main__":
    solution = Solution()
    
    test_cases = [
        (1, 1),
        (2, 2),
        (3, 3),
        (4, 5),
        (5, 8),
        (10, 89)
    ]
    
    for n, expected in test_cases:
        result = solution.climbStairs(n)
        print(f"Steps: {n} => Ways: {result} | Expected: {expected}")

File: Java/dynamic_programming/climbing_stairs.java

/*
    Problem: Climbing Stairs
    You are climbing a staircase. It takes n steps to reach the top.
    Each time you can climb either 1 step or 2 steps.
    Find how many distinct ways you can climb to the top.
*/

public class climbing_stairs {

    // -----------------------------------------------------------
    //  1. BRUTE FORCE (Recursive)
    // -----------------------------------------------------------
    // Idea:
    //   At each step, you can take 1 or 2 steps.
    //   So total ways = f(n-1) + f(n-2)
    //
    // Time Complexity: O(2^n)  (exponential)
    // Space Complexity: O(n)   (recursion call stack)
    public static int climbStairsBrute(int n) {
        if (n <= 2) return n; // base cases
        return climbStairsBrute(n - 1) + climbStairsBrute(n - 2);
    }


    // -----------------------------------------------------------
    //  2. OPTIMIZED DP (Bottom-Up Tabulation)
    // -----------------------------------------------------------
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    public static int climbStairsDP(int n) {
        if (n <= 2) return n;
        int[] dp = new int[n + 1];
        dp[1] = 1;
        dp[2] = 2;

        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }

        return dp[n];
    }


    // -----------------------------------------------------------
    //  3. OPTIMIZED (Space-Optimized DP)
    // -----------------------------------------------------------
    // We only need the last two results at any time.
    //
    // Time Complexity: O(n)
    // Space Complexity: O(1)
    public static int climbStairsOptimized(int n) {
        if (n <= 2) return n;
        int first = 1, second = 2;
        for (int i = 3; i <= n; i++) {
            int third = first + second;
            first = second;
            second = third;
        }
        return second;
    }


    // -----------------------------------------------------------
    //  MAIN METHOD (Test all versions)
    // -----------------------------------------------------------
    public static void main(String[] args) {
        int n = 5;  // example: 5 stairs

        System.out.println("Brute Force Result: " + climbStairsBrute(n));
        System.out.println("DP Result: " + climbStairsDP(n));
        System.out.println("Optimized Result: " + climbStairsOptimized(n));
    }
}
